state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,274 @@
+		Ok(())
+	}
+
+	fn is_vote_desired<ElectionAccess: ElectionReadAccess<ElectoralSystem = Self>>(
+		_election_identifier_with_extra: crate::electoral_system::ElectionIdentifierOf<Self>,
+		_election_access: &ElectionAccess,
+		_current_vote: Option<(VotePropertiesOf<Self>, AuthorityVoteOf<Self>)>,
+	) -> Result<bool, CorruptStorageError> {
+		Ok(true)
+	}
+
+	fn is_vote_needed(
+		(_, current_partial_vote, _): (
+			VotePropertiesOf<Self>,
+			<Self::Vote as VoteStorage>::PartialVote,
+			AuthorityVoteOf<Self>,
+		),
+		(proposed_partial_vote, _): (
+			<Self::Vote as VoteStorage>::PartialVote,
+			<Self::Vote as VoteStorage>::Vote,
+		),
+	) -> bool {
+		current_partial_vote != proposed_partial_vote
+	}
+
+	fn on_finalize<ElectoralAccess: ElectoralWriteAccess<ElectoralSystem = Self>>(
+		electoral_access: &mut ElectoralAccess,
+		election_identifiers: Vec<ElectionIdentifier<Self::ElectionIdentifierExtra>>,
+		current_block_number: &Self::OnFinalizeContext,
+	) -> Result<Self::OnFinalizeReturn, CorruptStorageError> {
+		if let Some(election_identifier) = election_identifiers
+			.into_iter()
+			.at_most_one()
+			.map_err(|_| CorruptStorageError::new())?
+		{
+			let mut election_access = electoral_access.election_mut(election_identifier)?;
+			if let Some(consensus) = election_access.check_consensus()?.has_consensus() {
+				let mut known_accounts = election_access.properties()?;
+				election_access.delete();
+				known_accounts.witnessed_open_accounts.extend(consensus.new_accounts.iter().map(
+					|(account, swap_details)| {
+						Hook::initiate_vault_swap((*swap_details).clone());
+						(*account).clone()
+					},
+				));
+				consensus.confirm_closed_accounts.into_iter().for_each(|acc| {
+					known_accounts.closure_initiated_accounts.remove(&acc);
+				});
+
+				if Hook::get_number_of_available_sol_nonce_accounts() >
+					NONCE_AVAILABILITY_THRESHOLD_FOR_INITIATING_SWAP_ACCOUNT_CLOSURES &&
+					(known_accounts.witnessed_open_accounts.len() >=
+						MAX_BATCH_SIZE_OF_VAULT_SWAP_ACCOUNT_CLOSURES ||
+						(*current_block_number)
+							.checked_sub(&electoral_access.unsynchronised_state()?)
+							.expect("current block number is always greater than when apicall was last created")
+							.into() >= MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS)
+				{
+					let accounts_to_close: Vec<_> = known_accounts
+					.witnessed_open_accounts
+					.drain(..sp_std::cmp::min(
+						known_accounts.witnessed_open_accounts.len(),
+						MAX_BATCH_SIZE_OF_VAULT_SWAP_ACCOUNT_CLOSURES
+					))
+					.collect();
+					match Hook::close_accounts(accounts_to_close.clone()) {
+						Ok(()) => {
+							known_accounts.closure_initiated_accounts.extend(accounts_to_close);
+							electoral_access.set_unsynchronised_state(*current_block_number)?;
+						},
+						Err(e) => {
+							log::error!(
+								"failed to close accounts: {:?}",
+								e
+							);
+							known_accounts.witnessed_open_accounts.extend(accounts_to_close);
```

Comment:
not really sure I understand why we do this?

Reply:
Ramiz can confirm but my understanding is that we are readding the `accounts_to_close` to the `known_accounts` in the scenario where the closing of accounts has failed. Since we had drained that account from the list we are readding it back.

Reply:
sorry yeah, didn't delete the comment... was more a comment to self when going through 👍